[CI] RollupActionIT testRollupIndexAndSetNewRollupPolicy failing

This is a new test case which has failed a few times now since being added.

**Build scan**:
https://gradle-enterprise.elastic.co/s/nckptifuyp4v2/tests/:x-pack:plugin:ilm:qa:multi-node:javaRestTest/org.elasticsearch.xpack.ilm.actions.RollupActionIT/testRollupIndexAndSetNewRollupPolicy#1

**Repro line**:
`./gradlew ':x-pack:plugin:ilm:qa:multi-node:javaRestTest' --tests "org.elasticsearch.xpack.ilm.actions.RollupActionIT.testRollupIndexAndSetNewRollupPolicy" -Dtests.seed=F0A4BF22F8036B7E -Dtests.security.manager=true -Dtests.locale=sk-SK -Dtests.timezone=Pacific/Truk -Druntime.java=11`

**Reproduces locally?**:
Nope.

**Applicable branches**:
`master` and `7.x` (error on the `7.x` build looked a bit different though)

**Failure history**:
https://gradle-enterprise.elastic.co/scans/tests?search.relativeStartTime=P7D&search.timeZoneId=America/Los_Angeles&tests.container=org.elasticsearch.xpack.ilm.actions.RollupActionIT&tests.sortField=FAILED&tests.unstableOnly=true

**Failure excerpt**:
```
org.elasticsearch.xpack.ilm.actions.RollupActionIT > testRollupIndexAndSetNewRollupPolicy FAILED
    java.lang.AssertionError
        at __randomizedtesting.SeedInfo.seed([F0A4BF22F8036B7E:7FF7996A3D79E9A]:0)
```

FYI, I've seen other builds with slightly different test failures as well but the common thing I see is this in some of the node logs:

```
» [2021-02-05T19:26:13,436][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [javaRestTest-0] fatal error in thread [elasticsearch[javaRestTest-0][rollup_indexing][T#1]], exiting
»  java.lang.AssertionError: null
»  	at org.elasticsearch.xpack.rollup.v2.RollupShardIndexer.checkCleanDirectory(RollupShardIndexer.java:207) ~[?:?]
»  	at org.elasticsearch.xpack.rollup.v2.RollupShardIndexer.execute(RollupShardIndexer.java:196) ~[?:?]
»  	at org.elasticsearch.xpack.rollup.v2.TransportRollupIndexerAction.shardOperation(TransportRollupIndexerAction.java:112) ~[?:?]
»  	at org.elasticsearch.xpack.rollup.v2.TransportRollupIndexerAction.shardOperation(TransportRollupIndexerAction.java:46) ~[?:?]
»  	at org.elasticsearch.action.support.broadcast.TransportBroadcastAction.lambda$asyncShardOperation$0(TransportBroadcastAction.java:291) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»  	at org.elasticsearch.action.ActionRunnable.lambda$supply$0(ActionRunnable.java:47) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»  	at org.elasticsearch.action.ActionRunnable$2.doRun(ActionRunnable.java:62) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»  	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:728) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»  	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»  	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
»  	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
»  	at java.lang.Thread.run(Thread.java:834) [?:?]
```

I think these might all be related.

Yeah, pretty much every build failing on task `:x-pack:plugin:ilm:qa:multi-node:javaRestTest` is giving me this on at least one of the nodes:

```
» [2021-02-05T14:52:12,154][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [javaRestTest-3] fatal error in thread [elasticsearch[javaRestTest-3][rollup_indexing][T#1]], exiting
»  java.lang.AssertionError: null
»  	at org.elasticsearch.xpack.rollup.v2.RollupShardIndexer.checkCleanDirectory(RollupShardIndexer.java:207) ~[?:?]
```

The confusing bit is depending on when it happens and which node the tests happen to hit you'll get different tests failing with different errors. I know very little, but it seems like that is the root cause here.

If we could prioritize a fix here that would be great since it's causing a whole host of build failures and it's not practical to mute them all unless we literally mute this entire project.

--- a/x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/actions/RollupActionIT.java
+++ b/x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/actions/RollupActionIT.java
@@ -61,7 +61,6 @@ public class RollupActionIT extends ESRestTestCase {
         assertBusy(() -> assertTrue(indexExists(index)));
     }
 
-    @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/68609")
     public void testRollupIndexAndSetNewRollupPolicy() throws Exception {
         createIndexWithSettings(client(), index, alias, Settings.builder().put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, 1)
             .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 0));

--- a/x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/RollupShardIndexer.java
+++ b/x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/RollupShardIndexer.java
@@ -98,11 +98,14 @@ class RollupShardIndexer {
     private final List<FieldValueFetcher> metricsFieldFetchers;
 
     private final CompressingOfflineSorter sorter;
-    private final Set<String> tmpFiles = new HashSet<>();
 
     private final BulkProcessor bulkProcessor;
     private final AtomicLong numSent = new AtomicLong();
     private final AtomicLong numIndexed = new AtomicLong();
+
+    // for testing
+    final Set<String> tmpFiles = new HashSet<>();
+    final Set<String> tmpFilesDeleted = new HashSet<>();
 
     RollupShardIndexer(Client client,
                        IndexService indexService,
@@ -130,6 +133,12 @@ class RollupShardIndexer {
                     IndexOutput output = super.createTempOutput(prefix, suffix, context);
                     tmpFiles.add(output.getName());
                     return output;
+                }
+
+                @Override
+                public void deleteFile(String name) throws IOException {
+                    tmpFilesDeleted.add(name);
+                    super.deleteFile(name);
                 }
             };
             this.searchExecutionContext = indexService.newSearchExecutionContext(
@@ -192,21 +201,10 @@ class RollupShardIndexer {
             do {
                 bucket = computeBucket(bucket);
             } while (bucket != null);
-        } finally {
-            assert(checkCleanDirectory(dir, tmpFiles));
         }
         // TODO: check that numIndexed == numSent, otherwise throw an exception
         logger.info("Successfully sent [" + numIndexed.get()  + "], indexed [" + numIndexed.get()  + "]");
         return numIndexed.get();
-    }
-
-    // check that all temporary files are deleted
-    private static boolean checkCleanDirectory(Directory dir, Set<String> tmpFiles) throws IOException {
-        Set<String> allFiles = Arrays.stream(dir.listAll()).collect(Collectors.toSet());
-        for (String tmpFile : tmpFiles) {
-            assert(allFiles.contains(tmpFile) == false);
-        }
-        return true;
     }
 
     private BulkProcessor createBulkProcessor() {

--- a/x-pack/plugin/rollup/src/test/java/org/elasticsearch/xpack/rollup/v2/RollupActionSingleNodeTests.java
+++ b/x-pack/plugin/rollup/src/test/java/org/elasticsearch/xpack/rollup/v2/RollupActionSingleNodeTests.java
@@ -19,7 +19,11 @@ import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.common.time.DateFormatter;
 import org.elasticsearch.common.xcontent.XContentBuilder;
 import org.elasticsearch.common.xcontent.XContentFactory;
+import org.elasticsearch.index.Index;
 import org.elasticsearch.index.IndexNotFoundException;
+import org.elasticsearch.index.IndexService;
+import org.elasticsearch.index.shard.IndexShard;
+import org.elasticsearch.indices.IndicesService;
 import org.elasticsearch.plugins.Plugin;
 import org.elasticsearch.search.aggregations.bucket.composite.CompositeAggregationBuilder;
 import org.elasticsearch.search.aggregations.bucket.composite.CompositeValuesSourceBuilder;
@@ -87,6 +91,31 @@ public class RollupActionSingleNodeTests extends ESSingleNodeTestCase {
                 "numeric_1", "type=double",
                 "numeric_2", "type=float",
                 "categorical_1", "type=keyword").get();
+    }
+
+    public void testRollupShardIndexerCleansTempFiles() throws IOException {
+        // create rollup config and index documents into source index
+        RollupActionDateHistogramGroupConfig dateHistogramGroupConfig = randomRollupActionDateHistogramGroupConfig("date_1");
+        SourceSupplier sourceSupplier = () -> XContentFactory.jsonBuilder().startObject()
+            .field("date_1", randomDateForInterval(dateHistogramGroupConfig.getInterval()))
+            .field("categorical_1", randomAlphaOfLength(1))
+            .field("numeric_1", randomDouble())
+            .endObject();
+        RollupActionConfig config = new RollupActionConfig(
+            new RollupActionGroupConfig(dateHistogramGroupConfig, null, new TermsGroupConfig("categorical_1")),
+            Collections.singletonList(new MetricConfig("numeric_1", Collections.singletonList("max"))));
+        bulkIndex(sourceSupplier);
+
+        IndicesService indexServices = getInstanceFromNode(IndicesService.class);
+        Index srcIndex = resolveIndex(index);
+        IndexService indexService = indexServices.indexServiceSafe(srcIndex);
+        IndexShard shard = indexService.getShard(0);
+
+        // re-use source index as temp index for test
+        RollupShardIndexer indexer = new RollupShardIndexer(client(), indexService, shard.shardId(), config, index, 2);
+        indexer.execute();
+        assertThat(indexer.tmpFilesDeleted, equalTo(indexer.tmpFiles));
+        // assert that files are deleted
     }
 
     public void testCannotRollupToExistingIndex() throws Exception {

